x-pack/plugins/infra/public/components/metrics_explorer/chart_context_menu.tsx

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+import React, { useCallback, useState } from 'react';
+import { i18n } from '@kbn/i18n';
+
+import {
+  EuiButtonEmpty,
+  EuiContextMenu,
+  EuiContextMenuPanelDescriptor,
+  EuiPopover,
+} from '@elastic/eui';
+import DateMath from '@elastic/datemath';
+import { Capabilities } from 'src/core/public';
+import { MetricsExplorerSeries } from '../../../common/http_api/metrics_explorer';
+import {
+  MetricsExplorerOptions,
+  MetricsExplorerTimeOptions,
+  MetricsExplorerChartOptions,
+} from '../../containers/metrics_explorer/use_metrics_explorer_options';
+import { createTSVBLink } from './helpers/create_tsvb_link';
+import { getNodeDetailUrl } from '../../pages/link_to/redirect_to_node_detail';
+import { SourceConfiguration } from '../../utils/source_configuration';
+import { InventoryItemType } from '../../../common/inventory_models/types';
+import { useKibana } from '../../../../../../src/plugins/kibana_react/public';
+
+interface Props {
+  options: MetricsExplorerOptions;
+  onFilter?: (query: string) => void;
+  series: MetricsExplorerSeries;
+  source?: SourceConfiguration;
+  timeRange: MetricsExplorerTimeOptions;
+  uiCapabilities?: Capabilities;
+  chartOptions: MetricsExplorerChartOptions;
+}
+
+const fieldToNodeType = (
+  source: SourceConfiguration,
+  field: string
+): InventoryItemType | undefined => {
+  if (source.fields.host === field) {
+    return 'host';
+  }
+  if (source.fields.pod === field) {
+    return 'pod';
+  }
+  if (source.fields.container === field) {
+    return 'container';
+  }
+};
+
+const dateMathExpressionToEpoch = (dateMathExpression: string, roundUp = false): number => {
+  const dateObj = DateMath.parse(dateMathExpression, { roundUp });
+  if (!dateObj) throw new Error(`"${dateMathExpression}" is not a valid time string`);
+  return dateObj.valueOf();
+};
+
+export const createNodeDetailLink = (
+  nodeType: InventoryItemType,
+  nodeId: string,
+  from: string,
+  to: string,
+  prefixPathWithBasePath: (app: string, path?: string | undefined) => string | undefined
+) => {
+  return getNodeDetailUrl({
+    nodeType,
+    nodeId,
+    from: dateMathExpressionToEpoch(from),
+    to: dateMathExpressionToEpoch(to, true),
+    prefixPathWithBasePath,
+  });
+};
+
+export const MetricsExplorerChartContextMenu = ({
+  onFilter,
+  options,
+  series,
+  source,
+  timeRange,
+  uiCapabilities,
+  chartOptions,
+}: Props) => {
+  const getUrlForApp = useKibana().services.application?.getUrlForApp;
+  const prependBasePath = useKibana().services.http?.basePath.prepend;
+  const prefixPathWithBasePath = useCallback(
+    (app: string, path?: string) => {
+      if (!getUrlForApp || !prependBasePath) {
+        return;
```

Comment:
Same as in the other context menu implementation, when is `undefined` a desirable return value?

Reply:
Will iron out with the other `prefixPathWithBasePath` changes.